extensions/grpc/deployment/src/main/java/io/quarkus/grpc/deployment/GrpcServerProcessor.java

```diff
@@ -212,41 +213,191 @@ void discoverBindableServices(BuildProducer<BindableServiceBuildItem> bindables,
+                    return annotationInstance != null && annotationInstance.target().kind() == Kind.METHOD;
+                };
+                BlockingMode methodMode = blockingMode(annotationOnMethod,
+                        () -> "Method '" + method.declaringClass().name() + "#" + method.name() +
+                                "' contains both @Blocking and @NonBlocking annotations.");
+                if (methodMode.explicit) {
+                    return methodMode;
                 }
-            } else if ((method.hasAnnotation(TRANSACTIONAL) || blockingClass) && !method.hasAnnotation(NON_BLOCKING)) {
-                result.add(method);
+                if (methodMode != BlockingMode.UNDEFINED) {
+                    mode = methodMode;
+                }
+            }
+        }
+        return mode;
+    }
+
+    private boolean effectiveBlockingMode(BlockingMode methodBlockingMode, BlockingMode serviceBlockingMode) {
+        if (methodBlockingMode.explicit) {
+            // explicit @Blocking or @NonBlocking on (any) method
+            return methodBlockingMode.blocking;
+        }
+        if (serviceBlockingMode.explicit) {
+            // explicit @Blocking or @NonBlocking on (any) class
+            return serviceBlockingMode.blocking;
+        }
+        // @Transactional on method or class
+        return methodBlockingMode.blocking || serviceBlockingMode.blocking;
+    }
+
+    /**
+     * Collect the names of all blocking methods.
+     *
+     * <p>
+     * Whether a method is blocking or not is evaluated for each individual service method (those that are defined
+     * in the generated {@code *ImplBase} class).
+     *
+     * <p>
+     * For each method:
+     * <ol>
+     * <li>blocking, if any of the method overrides has a {@link io.smallrye.common.annotation.Blocking} annotation.</li>
+     * <li>not-blocking, if any of the method overrides has a {@link io.smallrye.common.annotation.NonBlocking} annotation.</li>
+     * <li>blocking, if the service class or any of its base classes has a {@link io.smallrye.common.annotation.Blocking}
+     * annotation.</li>
+     * <li>non-blocking, if the service class or any of its base classes has a {@link io.smallrye.common.annotation.NonBlocking}
+     * annotation.</li>
+     * <li>blocking, if any of the method overrides has a {@link javax.transaction.Transaction} annotation.</li>
+     * <li>blocking, if the service class or any of its base classes has a {@link javax.transaction.Transaction}
+     * annotation.</li>
+     * <li>Else: non-blocking.</li>
+     * </ol>
+     */
+    private Set<String> gatherBlockingMethods(ClassInfo service, IndexView index) {
+        List<ClassInfo> classes = classHierarchy(service, index);
+
+        BlockingMode serviceBlockingMode = serviceBlockingMode(classes);
+
+        Set<String> result = new HashSet<>();
+
+        // Collect all gRPC methods from the *ImplBase class
```

Comment:
Again, an `*ImplBase` class does not have to be present.

Reply:
Updated the comment